apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SqlResultStream.cs

```diff
@@ -188,6 +194,14 @@ private async Task<PartialResultSet> ComputeNextAsync(CancellationToken cancella
                     _buffer.Clear();
                     _grpcCall.Dispose();
                     _grpcCall = null;
+
+                    // Calculate the remaining timeout.
```

Comment:
Can't all of this be moved to RetryState so we can just keep asking CanRetry as we are doing now?

Reply:
Maybe we can do something similar to what we have in [Gax](https://github.com/googleapis/gax-dotnet/blob/master/Google.Api.Gax.Grpc/ApiCallRetryExtensions.cs).

Reply:
(It's slightly fiddly though, in terms of whether CanRetry should check for timing. I'd be open to a bit of restructuring here - we don't want to duplicate checking in CanRetry and RecordErrorAndWaitAsync. Maybe CanRetry should work out how long the retry should be for as an out parameter, and then just call the scheduler delay directly in the finally block?)

Reply:
I've changed the implementation to using a deadline instead of calculating a remaining timeout. I'm not sure we should add the timing check to `CanRetry`:
1. If the deadline is exceeded, a `DeadlineExceeded` exception should be thrown and not the otherwise retryable error that occurred.
2. It does not seem logical to let `CanRetry` throw a `DeadlineExceeded` error if the error cannot be retried because it would exceed the deadline.
3. Letting `CanRetry` return false and a deadline that is in the past would not make this construct any simpler, as it would mean that the catch block would have to take that condition into account as well.

Another reason I would prefer not to have `CanRetry` return the delay as an output parameter is that it would not be deterministic because of the jittering of the retry delay (or we would have to delay the jittering until the actual wait is executed).